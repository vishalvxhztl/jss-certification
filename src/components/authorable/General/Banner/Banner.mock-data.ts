import { BannerProps } from './Banner';

// Realistically this is going to have a lot more data from Sitecore or XM Cloud, but this is a good example mock data set.
const defaultData: BannerProps = {
  rendering: { componentName: 'Banner' },
  params: {},
  fields: {
    contentPosition: {
      value: 'center', // Or left right
    },
    cta: {
      value: {
        href: 'https://www.horizontaldigital.com/',
        text: 'Learn more',
        linktype: 'external',
        url: 'https://www.horizontaldigital.com/',
        anchor: '',
        target: '_blank',
      },
    },
    description: {
      value:
        '<p>Join Horizontal at Connections 2023 and discover how we create personalized and connected experiences tailored for every customer.</p>',
    },
    image: {
      value: {
        src: 'https://dummyimage.com/16:9x1080/',
        alt: 'dummy Image',
      },
    },
    link: {
      value: {
        href: 'https://www.horizontaldigital.com/',
        text: 'Read more',
        linktype: 'external',
        url: 'https://www.horizontaldigital.com/',
        anchor: '',
        target: '_blank',
      },
    },
    title: {
      value: 'Connected experiences fine-tuned for your customers',
    },
  },
};

export const LeftBannerData: BannerProps = {
  ...defaultData,
  fields: {
    ...defaultData.fields,
    contentPosition: { ...defaultData.fields.contentPosition, value: 'left' },
  },
};

export const RightBannerData: BannerProps = {
  ...defaultData,
  fields: {
    ...defaultData.fields,
    contentPosition: { ...defaultData.fields.contentPosition, value: 'right' },
  },
};

export const noData = {
  rnder: {},
  params: [],
};

export default defaultData;
