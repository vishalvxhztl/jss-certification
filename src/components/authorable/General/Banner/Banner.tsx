// import { Field, withDatasourceCheck } from '@sitecore-jss/sitecore-jss-nextjs';
import ImageHelper from '@/components/helpers/ImageHelper/ImageHelper';
import LinkWrapper from '@/components/helpers/LinkWrapper/LinkWrapper';
import Text from '@/components/helpers/Text/Text';
import { Field, ImageField, LinkField, RichTextField } from '@sitecore-jss/sitecore-jss-nextjs';
import clsx from 'clsx';

// Local
import RichTextA11yWrapper from 'components/helpers/RichTextA11yWrapper/RichTextA11yWrapper';

// Ideally, all this is from generated Typescript code from Sitecore and we're not manually defining types.
interface BannerFields {
  contentPosition: Field<string>;
  cta: LinkField;
  description: RichTextField;
  image: ImageField;
  link: LinkField;
  title: Field<string>;
}

export type BannerProps = {
  rendering: { componentName: string };
  params: { [key: string]: string };
  fields: BannerFields;
};

const Banner = ({ fields }: BannerProps): JSX.Element => {
  //Handle Classname as per sitecore position
  const alignmentClass =
    fields?.contentPosition?.value.trim().toLowerCase() === 'left'
      ? 'justify-start'
      : fields?.contentPosition?.value.trim().toLowerCase() === 'right'
      ? 'justify-end'
      : 'justify-center';
  // console.log('Banner || alignmentClass :', alignmentClass);

  // Fail out if fields aren't present
  if (fields === null || fields === undefined) return <></>;

  return (
    <div
      className={clsx(
        'bg-theme-bg',
        'relative',
        'h-screen',
        'max-h-full',
        'w-full',
        'flex',
        alignmentClass
      )}
      data-component="authorable/general/banner"
      data-testid="Banner"
    >
      {/* Image Section */}
      <ImageHelper className={clsx('object-cover', 'z-0')} image={fields?.image} fill={true} />
      {/* Content Section */}
      <div
        className={clsx(
          'z-1',
          'absolute',
          'top-1/2',
          '-translate-y-1/2',
          'md:w-1/3',
          'flex',
          'flex-col',
          'items-center',
          'gap-2',
          'p-5',
          'm-2',
          'bg-black',
          'bg-opacity-20',
          'rounded-lg',
          'backdrop-blur-sm',
          'shadow-sm'
        )}
      >
        <Text
          text={fields?.title?.value}
          tag="h2"
          className={clsx('font-bold', 'text-center', 'text-lg', 'md:text-2xl', 'lg:text-3xl')}
        />
        <RichTextA11yWrapper field={fields?.description} className={clsx('text-center')} editable />
        <LinkWrapper
          field={fields?.link}
          className={clsx('text-blue-600', 'hover:text-blue-500')}
        />
        <LinkWrapper field={fields?.cta} className={clsx('btn-full')} />
      </div>
    </div>
  );
};

// @todo: Figure out how to mock isPageEditing, or if it even matters, in Storybook.
// export default withDatasourceCheck()<ContentBlockProps>(ContentBlock);
export default Banner;
