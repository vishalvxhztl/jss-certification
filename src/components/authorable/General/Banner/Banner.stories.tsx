// Global
import type { Meta, StoryObj } from '@storybook/react';

// Lib
import { expandObj, flattenObj } from 'lib/object-parser';

// Local
import Banner, { BannerProps } from './Banner';
import defaultData, { LeftBannerData, RightBannerData } from './Banner.mock-data';

const meta: Meta<typeof Banner> = {
  title: 'Authorable/General/Banner',
  component: Banner,
  argTypes: {},
};

export default meta;

type Story = StoryObj<typeof Banner>;

export const Default: Story = {
  render: (args) => {
    return <Banner {...(expandObj({ ...args }) as BannerProps)} />;
  },
  args: {
    ...flattenObj(defaultData),
  },
};

export const Left: Story = {
  render: (args) => {
    return (
      <Banner
        {...(expandObj({
          ...args,
        }) as BannerProps)}
      />
    );
  },
  args: {
    ...flattenObj(LeftBannerData),
  },
};
export const Right: Story = {
  render: (args) => {
    return (
      <Banner
        {...(expandObj({
          ...args,
        }) as BannerProps)}
      />
    );
  },
  args: {
    ...flattenObj(RightBannerData),
  },
};
