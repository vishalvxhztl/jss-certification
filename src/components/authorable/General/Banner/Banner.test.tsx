// Lib
import { snapshot, hasDataComponent, renderComponent } from 'lib/testing/utils';

// Local
import Banner from './Banner';
import defaultData, { noData } from './Banner.mock-data';

it('renders correctly', () => {
  const component = snapshot(Banner, { componentProps: defaultData });
  hasDataComponent(component, 'authorable/general/banner');
});

it('does not render without data', () => {
  const component = renderComponent(Banner, { componentProps: noData });
  const text = component.queryByTestId('Banner');
  expect(text).toBe(null);
});
