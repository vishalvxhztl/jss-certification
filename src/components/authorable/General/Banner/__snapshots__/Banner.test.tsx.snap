// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  class="bg-theme-bg relative h-screen max-h-full w-full flex justify-center"
  data-component="authorable/general/banner"
  data-testid="Banner"
>
  <img
    alt="dummy Image"
    class="object-cover z-0"
    data-nimg="fill"
    decoding="async"
    loading="lazy"
    sizes="100vw"
    src="/_next/image?url=https%3A%2F%2Fdummyimage.com%2F16%3A9x1080%2F&w=3840&q=75"
    srcset="/_next/image?url=https%3A%2F%2Fdummyimage.com%2F16%3A9x1080%2F&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fdummyimage.com%2F16%3A9x1080%2F&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fdummyimage.com%2F16%3A9x1080%2F&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fdummyimage.com%2F16%3A9x1080%2F&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fdummyimage.com%2F16%3A9x1080%2F&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fdummyimage.com%2F16%3A9x1080%2F&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fdummyimage.com%2F16%3A9x1080%2F&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fdummyimage.com%2F16%3A9x1080%2F&w=3840&q=75 3840w"
    style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
  />
  <div
    class="z-1 absolute top-1/2 -translate-y-1/2 md:w-1/3 flex flex-col items-center gap-2 p-5"
  >
    <h2
      class="font-bold text-center text-lg md:text-2xl lg:text-3xl"
    >
      Connected experiences fine-tuned for your customers
    </h2>
    <div
      class="rte"
    >
      <p>
        Join Horizontal at Connections 2023 and discover how we create personalized and connected experiences tailored for every customer.
      </p>
    </div>
    <a
      class="text-blue-500 hover:text-blue-800"
      href="https://www.horizontaldigital.com/"
      rel="noopener noreferrer"
      target="_blank"
    >
      Read more
      <span
        class="sr-only"
      >
         (Opens in a new tab)
      </span>
      <span />
    </a>
    <a
      class="btn-full"
      href="https://www.horizontaldigital.com/"
      rel="noopener noreferrer"
      target="_blank"
    >
      Learn more
      <span
        class="sr-only"
      >
         (Opens in a new tab)
      </span>
      <span />
    </a>
  </div>
</div>
`;
